### PR TITLE
[FIX] loyalty, pos_loyalty: download, print, mail gift card

### PR DESCRIPTION
**Pull request by trj-odoo**
- Gift cards can now be sent alongside the receipt via mail in the Receipt Screen.
- The gift card report no longer show n² gift cards
- Properly download all PDF files for the gift cards when there are multiple gift card programs

### CODE REVIEW

File: addons/pos_loyalty/models/pos_config.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ def _check_before_creating_new_session(self):
         if invalid_reward_products_msg:
             prefix_error_msg = _("To continue, make the following reward products available in Point of Sale.")
             raise UserError(f"{prefix_error_msg}\n{invalid_reward_products_msg}")
-        if  gift_card_programs:
```
Comment (pimodoo) [inline addons/pos_loyalty/models/pos_config.py:53]: useless diff?

File: addons/loyalty/report/loyalty_report_templates.xml [hunk 2]
```diff
@@ -75,36 +75,34 @@
 
     <template id="gift_card_report">
         <t t-call="web.html_container">
-            <t t-foreach="docs" t-as="o">
```
Comment (pimodoo) [inline addons/loyalty/report/loyalty_report_templates.xml:78]: why did you need to remove that?

File: addons/pos_loyalty/models/pos_config.py (hunk 1 above)
Comment (trj-odoo) [reply addons/pos_loyalty/models/pos_config.py:53]: Kind of, I saw two space instead of one and it triggered me

File: addons/pos_loyalty/models/pos_order.py [hunk 3]
```diff
@@ -146,3 +147,29 @@ def _get_fields_for_order_line(self):
...
+        attachment = super()._add_mail_attachment(name, ticket)
+        gift_card_programs = self.config_id._get_program_ids().filtered(lambda p: p.program_type == 'gift_card')
+        if gift_card_programs:
+            gift_cards = self.env['loyalty.card'].search([('source_pos_order_id', '=', self.id),
+                                                          ('program_id', 'in', gift_card_programs.mapped('id'))])
+            if gift_cards:
+                for program in gift_card_programs:
+                    filtered_gift_cards = gift_cards.filtered(lambda gc: gc.program_id == program)
+                    if filtered_gift_cards:
+                        action_report = program.pos_report_print_id
```
Comment (pimodoo) [inline addons/pos_loyalty/models/pos_order.py:161]: shouldn't we check that there is a 'pos_report_print_id' before using it?

File: addons/loyalty/report/loyalty_report_templates.xml (hunk 2 above)
Comment (trj-odoo) [reply addons/loyalty/report/loyalty_report_templates.xml:78]: because this template is still called in a `foreach` loop for every gift card in `gift_card_report_i18n`, this ends up creating a bug where if you have 2 gift cards, it would print 4 (each card twice), if you have 3, it would print 9, etc

File: addons/pos_loyalty/models/pos_order.py (hunk 3 above)
Comment (trj-odoo) [reply addons/pos_loyalty/models/pos_order.py:161]: Robin made a fix where this become mandatory https://github.com/odoo/odoo/pull/102613

Comment (trj-odoo) [reply addons/pos_loyalty/models/pos_order.py:161]: oh nvm, I still need to check it since the settings can be on `scan_use`

Comment (pimodoo) [reply addons/pos_loyalty/models/pos_order.py:161]: only in a specific configuration ( if self.gift_card_settings == "create_set":). If you use an existing card, you have no template.

Comment (trj-odoo) [reply addons/pos_loyalty/models/pos_order.py:162]: done in line 154

**Comment by pimodoo** [comment]
@robodoo r+ rebase-ff
